xbmc/threads/Thread.cpp

```diff
@@ -102,6 +102,7 @@ void CThread::Create(bool bAutoDelete, unsigned stacksize)
 
 bool CThread::IsRunning() const
 {
+  CSingleLock lock(m_CriticalSection);
```

Comment:
>No, it is not undefined behaviour

I recommend you to read http://en.cppreference.com/w/cpp/language/memory_model Specifically
"When an evaluation of an expression writes to a memory location and another evaluation reads or modifies the same memory location, the expressions are said to conflict. A program that has two conflicting evaluations has a data race unless either

both conflicting evaluations are atomic operations (see std::atomic)
one of the conflicting evaluations happens-before another (see std::memory_order) 

If a data race occurs, the behavior of the program is undefined.
(in particular, release of a std::mutex is synchronized-with, and therefore, happens-before acquisition of the same mutex by another thread, which makes it possible to use mutex locks to guard against data races) "

Thus, the way I see it, scenario 1:
T1: executes pThread->m_ThreadId = 0;
T2: executes pThread->m_ThreadId;  =undefined behaviour

scenario 2: change staticThread to this:
````
pThread->m_StartEvent.Set();
//use m_ThreadId store some temporary here that happens to be 0. to any other threads reading it it will be 0 but it's sequentially consistent and a perfectly valid optimization.
pThread->m_ThreadId = 0;
pThread->Action();
// lock during termination
CSingleLock lock(pThread->m_CriticalSection);
pThread->m_ThreadId = 0;
````
adding a lock or making it atomic ensures scenario 1 & 2 can't happen.